tests/js/searchtools.js

```diff
@@ -24,34 +22,34 @@ describe('Basic html theme search', function() {
         5,
         "index.rst"
       ]];
-      expect(Search.performTermsSearch(searchterms, excluded, terms, titleterms)).toEqual(hits);
+      expect(Search._performSearch(...searchTerms)).toEqual(hits);
     });
 
     it('should be able to search for multiple terms', function() {
-      index = {
-        alltitles: {
-          'Main Page': [[0, 'main-page']],
-        },
-        docnames:["index"],
-        filenames:["index.rst"],
-        terms:{main:0, page:0},
-        titles:["Main Page"],
-        titleterms:{ main:0, page:0 }
-      }
-      Search.setIndex(index);
-
-      searchterms = ['main', 'page'];
-      excluded = [];
-      terms = index.terms;
-      titleterms = index.titleterms;
-      hits = [[
-        'index',
-        'Main Page',
-        '',
-        null,
-        15,
-        'index.rst']];
-      expect(Search.performTermsSearch(searchterms, excluded, terms, titleterms)).toEqual(hits);
+      eval(loadFixture("multiterm/searchindex.js"));
```

Comment:
I don't understand why the two tests are now different.

Reply:
The closest analogy for this that I can think of is the `srcdir=...` setting for the Python tests.  Each of the JavaScript test cases can (and in the case of the search tests, does) now read from test fixture(s) relevant to that individual test case.

It might be possible to refactor this so that the `loadFixture` happens implicitly / during some kind of `setup` phase based on the test name.

Reply:
Actually, I don't understand why the duplicated are found with the fixture but not with the previous test (was there anything that changed? I thought that you just moved the hard-coded fixtures in a separate fixture file).

Reply:
> It might be possible to refactor this so that the `loadFixture` happens implicitly / during some kind of `setup` phase based on the test name.

The relevant `jasmine` guidance to follow if/when doing this appears to be: https://jasmine.github.io/tutorials/sharing_behaviors#sharing-behaviors

Reply:
> Actually, I don't understand why the duplicated are found with the fixture but not with the previous test (was there anything that changed? I thought that you just moved the hard-coded fixtures in a separate fixture file).

Good question.  Checking why that happened..

Reply:
Ah yes, I forgot about this.  The old test code was only testing a portion of the search query handling - the `performTermsSearch` function.

The duplicate is found during another part of the search code, namely here: https://github.com/sphinx-doc/sphinx/blob/600d3d3adaee530363c88ab1a000ef83e5db428e/sphinx/themes/basic/static/searchtools.js#L323-L338

(note when navigating further down that code file that [`performTermsSearch` is called later](https://github.com/sphinx-doc/sphinx/blob/600d3d3adaee530363c88ab1a000ef83e5db428e/sphinx/themes/basic/static/searchtools.js#L367-L368))

When we update the code here to use the top-level `_parseQuery` function -- that encompasses all of the query-handling code -- the duplicate appears in the results because we check both `term` and `title` matches.

Reply:
Could you make the old test compatible with your fixtures as well? it'd be good to test specific components.

Reply:
To check I understand you: partially revert the changes to the `should be able to search for multiple terms`, so that it continues to load a search index from fixtures, and only tests the `performTermsSearch` function, as-before.

Reply:
With the testroot in Python you have E2E but sometimes you only want to check a single component's logic so I would like to keep single component logic testing separate. In other words:

- keep the old test (manually creating the fixture is fine); it's just a logic unit being tested.
- you create a new test for testing what you just added (performing the full E2E, with the duplicates).

Reply:
Makes sense - I'm thinking about it; doing that would mean that the two tests duplicate a lot of the coverage and behaviour that they're testing.

I'm also wondering whether to extract the query refactoring into a separate smaller pull request, without the test fixture changes.

Reply:
> Makes sense - I'm thinking about it; doing that would mean that the two tests duplicate a lot of the coverage and behaviour that they're testing.
> 
> I'm also wondering whether to extract the query refactoring into a separate smaller pull request, without the test fixture changes.

Yep, let's do that first, if that's OK?  I've opened #12229 for it.

Reply:
> > With the testroot in Python you have E2E but sometimes you only want to check a single component's logic so I would like to keep single component logic testing separate. In other words:
> >
> >  * keep the old test (manually creating the fixture is fine); it's just a logic unit being tested.
> >
> >  * you create a new test for testing what you just added (performing the full E2E, with the duplicates).
> 
> Makes sense - I'm thinking about it; doing that would mean that the two tests duplicate a lot of the coverage and behaviour that they're testing.

Ok, agreed - I think it is OK to have both types of test in place, and I've pushed corresponding changes.